src/backend/distributed/commands/dependencies.c

```diff
@@ -88,6 +88,16 @@ EnsureDependenciesExistOnAllNodes(const ObjectAddress *target)
 	 */
 	List *workerNodeList = ActivePrimaryNonCoordinatorNodeList(RowShareLock);
 
+	/*
+	 * Lock dependent objects explicitly to make sure same DDL command won't be sent
+	 * multiple times from parallel sessions.
+	 */
+	foreach_ptr(dependency, dependenciesWithCommands)
+	{
+		LockDatabaseObject(dependency->classId, dependency->objectId,
```

Comment:
Two concurrent `RowExclusiveLock` would not block each other. Is this really what we want? See [here](https://www.postgresql.org/docs/14/explicit-locking.html) for the details.

I was expecting to see `ExclusiveLock`. Did you try if this fixes your problem in #5415?

Reply:
Yes, you are right. ExclusiveLock also fixed my problem on #5415.

Reply:
Are there any deadlock risks here?

Reply:
I initially thought that because dependencies are already sorted, there is no risk for the same object's distribution. 

But, there is a small risk if concurrent operations distribute different objects, where two objects are created in different order, hence deadlock. I'm not sure if that could happen in practice though.

One idea could be to sort objects based on objectAddress before acquiring the locks so we avoid this situation altogether.

Reply:
I've sorted the dependencies which will be created on workers by their address. Since we've ON CONFLICT DO NOTHING on MarkObjectDistributed and DDL commands are idempotent on workers, I don't recheck the dependencies after locking them.